### PR TITLE
Return x.alias() when transpose is an nop

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84674

To fix bug in https://gist.github.com/SherlockNoMad/b8dfbc614d3e65707d1bc379a098196d 

```
def f(x):
    return x.t()

x = torch.randn(2, requires_grad=True)
y = f(x)

compiled_f = make_fx(f)(x)
y_compiled = compiled_f(x)


print(compiled_f)
print("y.requires_grad", y.requires_grad)
print("y_compiled.requires_grad", y_compiled.requires_grad)


# def forward(self, x_1):
#     t_default = torch.ops.aten.t.default(x_1);  x_1 = None
#     detach_default = torch.ops.aten.detach.default(t_default);  t_default = None
#     return detach_default

# y.requires_grad True
# y_compiled.requires_grad False
```

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -2532,7 +2532,7 @@ Tensor transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
 
   // Transpose of a tensor is a view operation.
   if (dim0 == dim1) {
-    return self;
+    return self.alias();
```
Comment (SherlockNoMad) [inline aten/src/ATen/native/TensorShape.cpp:2535]: This is breaking the SparseCSR test 

I am special case this for SparseCSR Tensor

    if (self.is_sparse_csr()) {
      return self.clone();
    } else {
      return self.alias();
    }

Comment (Chillee) [reply aten/src/ATen/native/TensorShape.cpp:2535]: Seems right, according to the other behavior in the file.

Comment (cpuhrsch) [reply aten/src/ATen/native/TensorShape.cpp]: Please don't incur a copy when it's actually not needed.

Comment (Chillee) [reply aten/src/ATen/native/TensorShape.cpp]: I assumed that was what this was doing? https://github.com/pytorch/pytorch/pull/84674/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cfR2523

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -3216,7 +3216,19 @@ Tensor view(const Tensor& self,
 }
 
 Tensor alias(const Tensor& self) {
-    return alias_with_sizes_and_strides(self, self.sizes(), self.strides());
+  if (self.is_sparse_csr()) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorShape.cpp:3219]: Thanks for adding this! You can expand this to more sparse formats by using `self.is_sparse_csc`, `self.is_sparse_bsr`, `self.is_sparse_bsc`. Can you also update our unit tests for alias to try these layouts? You can construct sparse tensors via `to_sparse_csr` etc. from 2D dense Tensors.

Comment (SherlockNoMad) [reply aten/src/ATen/native/TensorShape.cpp:3219]: test_transpose in test_sprase_csr.py is already testing with all_sparse_compressed_layouts.
Does it imply other sparse format is supporting regular .alias() call ? 

```
    @all_sparse_compressed_layouts()
    def test_transpose(self, device, dtype, layout):
```

Comment (cpuhrsch) [reply aten/src/ATen/native/TensorShape.cpp:3219]: @SherlockNoMad - I thought we have standalone tests for alias that need to be extended with support for these additional layouts.

Comment (SherlockNoMad) [reply aten/src/ATen/native/TensorShape.cpp:3219]: After rebasing to the latest master, I found that a recent PR already handled sparse tensors with sparse_compressed_transpose function. 
So this special case for alias is no longer needed. 

I am removing this now.

Comment (SherlockNoMad) [reply aten/src/ATen/native/TensorShape.cpp:3219]: Handled in this PR https://github.com/pytorch/pytorch/pull/82122

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**